src/wifi_manager.c

```diff
@@ -616,83 +648,356 @@ wifi_manager_destroy()
+        if (wifi_manager_lock_json_buffer(portMAX_DELAY))
         {
-            memcpy(first_free, ap, sizeof(wifi_ap_record_t));
-            memset(ap, 0, sizeof(wifi_ap_record_t));
-            /* find the next free slot */
-            for (int j = 0; j < *aps; j++)
-            {
-                if (aplist[j].ssid[0] == 0)
-                {
-                    first_free = &aplist[j];
-                    break;
-                }
-            }
+            wifi_manager_generate_ip_info_json(UPDATE_LOST_CONNECTION);
+            wifi_manager_unlock_json_buffer();
         }
+
+        wifiman_msg_send_cmd_connect_sta(CONNECTION_REQUEST_AUTO_RECONNECT);
     }
-    /* update the length of the list */
-    *aps = total_unique;
 }
 
-void
-wifi_manager_set_callback(message_code_t message_code, wifi_manager_cb_ptr func_ptr)
+static void
+wifi_handle_cmd_start_ip(void)
 {
-    if (message_code < MESSAGE_CODE_COUNT)
+    ESP_LOGI(TAG, "MESSAGE: ORDER_START_AP");
+    esp_wifi_set_mode(WIFI_MODE_APSTA);
+
+    // http_server_start();
+    // dns_server_start();
+}
+
+static void
+wifi_handle_ev_got_ip(const wifiman_msg_param_t *p_param)
+{
+    ESP_LOGI(TAG, "MESSAGE: EVENT_STA_GOT_IP");
+
+    const EventBits_t uxBits = xEventGroupGetBits(wifi_manager_event_group);
+
+    /* reset connection requests bits -- doesn't matter if it was set or not */
+    xEventGroupClearBits(wifi_manager_event_group, WIFI_MANAGER_REQUEST_STA_CONNECT_BIT);
+
+    /* save IP as a string for the HTTP server host */
+    const sta_ip_address_t ip_addr = wifiman_conv_param_to_ip_addr(p_param);
+    sta_ip_safe_set(ip_addr, portMAX_DELAY);
+
+    /* save wifi config in NVS if it wasn't a restored of a connection */
+    if (0 != (uxBits & (uint32_t)WIFI_MANAGER_REQUEST_RESTORE_STA_BIT))
     {
-        cb_ptr_arr[message_code] = func_ptr;
+        xEventGroupClearBits(wifi_manager_event_group, WIFI_MANAGER_REQUEST_RESTORE_STA_BIT);
+    }
+    else
+    {
+        wifi_manager_save_sta_config();
     }
+
+    /* refresh JSON with the new IP */
+    if (wifi_manager_lock_json_buffer(portMAX_DELAY))
+    {
+        /* generate the connection info with success */
+        wifi_manager_generate_ip_info_json(UPDATE_CONNECTION_OK);
+        wifi_manager_unlock_json_buffer();
+    }
+    else
+    {
+        abort();
+    }
+
+    /* bring down DNS hijack */
+    // http_server_stop();
+    dns_server_stop();
 }
 
-void
-wifi_manager(void *pvParameters)
+static void
+wifi_handle_cmd_disconnect_sta(void)
+{
+    ESP_LOGI(TAG, "MESSAGE: ORDER_DISCONNECT_STA");
+
+    /* precise this is coming from a user request */
+    xEventGroupSetBits(wifi_manager_event_group, WIFI_MANAGER_REQUEST_DISCONNECT_BIT);
+
+    /* order wifi discconect */
+    ESP_ERROR_CHECK(esp_wifi_disconnect());
+}
+
+ATTR_NORETURN
+static void
+wifi_manager_main_loop(void)
+{
+    for (;;)
+    {
+        queue_message msg = { 0 };
+        if (!wifiman_msg_recv(&msg))
+        {
+            continue;
```

Comment:
Should here be some kind of a delay/sleep? This looks like it would take 100% of available CPU time.

Reply:
fixed, comment added.